### PR TITLE
`Top` & `Rare` PPL commands support

### LINKED ISSUE #461: [FEATURE] add `Rare` PPL command
Add PPL `rare` command described [here](https://github.com/opensearch-project/sql/blob/7e73f124e2dc63a69750c8034558d6870c46ccf9/docs/user/ppl/cmd/rare.rst)

**Do you have any additional context?**
 - https://github.com/opensearch-project/opensearch-spark/issues/408

### PR DESCRIPTION
**Pull request by YANG-DB**
### Description
Adding support for Rare & Top PPL 

`top [N] <field-list> [by-clause]`

> N: number of results to return. Default: 10
> field-list: mandatory. comma-delimited list of field names.
> by-clause: optional. one or more fields to group the results by.
-------------------------------------------------------------------------------------------

`rare <field-list> [by-clause]`

>field-list: mandatory. comma-delimited list of field names.
>by-clause: optional. one or more fields to group the results by.
-------------------------------------------------------------------------------------------
**commands:**
 - https://github.com/opensearch-project/opensearch-spark/issues/461
 - https://github.com/opensearch-project/opensearch-spark/issues/536

### Issues Resolved
#461
#536 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: integ-test/src/integration/scala/org/opensearch/flint/spark/ppl/FlintSparkPPLTopAndRareITSuite.scala [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    val sortedPlan: LogicalPlan =
+      Sort(
+        Seq(SortOrder(UnresolvedAttribute("address"), Descending)),
+        global = true,
+        aggregatePlan)
+    val expectedPlan = Project(projectList, sortedPlan)
+    comparePlans(expectedPlan, logicalPlan, false)
+  }
+
+  test("create ppl rare address by age field query test") {
```
Comment (vamsi-amazon) [inline integ-test/src/integration/scala/org/opensearch/flint/spark/ppl/FlintSparkPPLTopAndRareITSuite.scala:80]: Tangential question: why does rare not have number of results like top?

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java [hunk 2]
```diff
@@ -174,20 +179,39 @@ public LogicalPlan visitAggregation(Aggregation node, CatalystPlanContext contex
...
             span.accept(this, context);
             //add span's group alias field (most recent added expression)
             context.getGroupingParseExpressions().add(context.getNamedParseExpressions().peek());
         }
         // build the aggregation logical step
-        return extractedAggregation(context);
+        LogicalPlan logicalPlan = extractedAggregation(context);
+
+        // set sort direction according to command type (`rare` is Asc, `top` is Desc, default to Asc)
+        List<SortDirection> sortDirections = new ArrayList<>();
```
Comment (vamsi-amazon) [inline ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:197]: Why is this a list? When we are only using the zeroth element.

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java [hunk 3]
```diff
@@ -174,20 +179,39 @@ public LogicalPlan visitAggregation(Aggregation node, CatalystPlanContext contex
...
+        sortDirections.add(node instanceof RareAggregation ? Descending$.MODULE$ : Ascending$.MODULE$);
+
+        if (!node.getSortExprList().isEmpty()) {
```
Comment (vamsi-amazon) [inline ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:200]: Nit: Can we break this into new methods?

File: ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanAggregationQueriesTranslatorTestSuite.scala [hunk 4]
```diff
@@ -42,6 +42,55 @@ class PPLLogicalPlanAggregationQueriesTranslatorTestSuite
     comparePlans(expectedPlan, logPlan, false)
   }
 
+  test("test count price") {
```
Comment (vamsi-amazon) [inline ppl-spark-integration/src/test/scala/org/opensearch/flint/spark/ppl/PPLLogicalPlanAggregationQueriesTranslatorTestSuite.scala:45]: Are the below two test cases related to the current PR?

**Comment by vamsi-amazon** [comment]
LGTM added minor comments.

File: integ-test/src/integration/scala/org/opensearch/flint/spark/ppl/FlintSparkPPLTopAndRareITSuite.scala (hunk 1 above)
Comment (YANG-DB) [reply integ-test/src/integration/scala/org/opensearch/flint/spark/ppl/FlintSparkPPLTopAndRareITSuite.scala:80]: rare has default 10 results similar to the definition [here](https://github.com/opensearch-project/sql/blob/main/docs/user/ppl/cmd/rare.rst#description)

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java (hunk 2 above)
Comment (YANG-DB) [reply ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:197]: yes - this is due to java limitation for changing an object which is not final in a lambda expression - list if a way to override this - would be happy for another alternative ?

File: ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java (hunk 3 above)
Comment (YANG-DB) [reply ppl-spark-integration/src/main/java/org/opensearch/sql/ppl/CatalystQueryPlanVisitor.java:200]: I think we need to review the entire `CatalystQueryPlanVisitor.java` class and refactor it as a command pattern only to contain visitor api and dispatch execution to a dedicated command class - (double dispatch)
lets do this but in another PR ...

**Pull request merged**